File: backend/rag.py
"""
Retrieval-Augmented Generation pipeline for the RTI portal.

Indexes the ministry's mock data — past responded RTI requests (precedent) and
FAQs (process knowledge) — so the AI step can ground its drafts in concrete
prior work, alongside the live web sources.

Architecture:
- `Embedder` is a Protocol — production uses `SentenceTransformersEmbedder`
  with `all-MiniLM-L6-v2`; tests use a deterministic stub.
- `RAGIndex` holds the embedding matrix in memory (numpy). Cosine similarity
  is fine at our scale (≤ a few hundred items).
- Newly approved requests get added to the index via `add_request_to_index`
  so each officer approval enriches the retrievable corpus.
"""

from __future__ import annotations

import logging
import threading
from typing import Optional, Protocol

import numpy as np

log = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Embedder
# ---------------------------------------------------------------------------


class Embedder(Protocol):
    def embed(self, texts: list[str]) -> np.ndarray:
        """Return an array of shape (len(texts), dim) of L2-normalized vectors."""
        ...


class SentenceTransformersEmbedder:
    """Lazy-loaded singleton wrapper around sentence-transformers."""

    _MODEL_NAME = "sentence-transformers/all-MiniLM-L6-v2"

    def __init__(self) -> None:
        self._model = None
        self._lock = threading.Lock()

    def _load(self):
        if self._model is not None:
            return self._model
        with self._lock:
            if self._model is None:
                from sentence_transformers import SentenceTransformer  # local import keeps tests fast
                log.info("Loading sentence-transformers model %s ...", self._MODEL_NAME)
                self._model = SentenceTransformer(self._MODEL_NAME)
        return self._model

    def embed(self, texts: list[str]) -> np.ndarray:
        if not texts:
            return np.zeros((0, 384), dtype=np.float32)
        model = self._load()
        vectors = model.encode(
            texts,
            normalize_embeddings=True,
            convert_to_numpy=True,
            show_progress_bar=False,
        )
        return vectors.astype(np.float32)


# ---------------------------------------------------------------------------
# Index
# ---------------------------------------------------------------------------


def _request_text(req: dict) -> str:
    """Searchable text for an RTI request: subject + description (+ response if responded)."""
    parts = [
        req.get("subject", ""),
        req.get("description", ""),
    ]
    response = req.get("response")
    if response and req.get("status") in {"Responded", "Under Review"}:
        parts.append(response)
    return "\n".join(p for p in parts if p)


def _faq_text(faq: dict) -> str:
    return f"{faq.get('question', '')}\n{faq.get('answer', '')}"


class RAGIndex:
    """
    Single in-memory index keyed by string id. Embedding matrix is a numpy
    float32 array of shape (N, dim); ids and payloads are parallel lists.
    Cosine similarity is dot product because all vectors are L2-normalized.
    """

    def __init__(self, embedder: Embedder) -> None:
        self._embedder = embedder
        self._ids: list[str] = []
        self._payloads: list[dict] = []
        self._matrix: Optional[np.ndarray] = None
        self._lock = threading.Lock()

    def __len__(self) -> int:
        return len(self._ids)

    # --- write path -------------------------------------------------------

    def reset(self) -> None:
        with self._lock:
            self._ids.clear()
            self._payloads.clear()
            self._matrix = None

    def upsert(self, item_id: str, text: str, payload: dict) -> None:
        """Add or replace an item by id."""
        vec = self._embedder.embed([text])  # (1, dim)
        with self._lock:
            if item_id in self._ids:
                idx = self._ids.index(item_id)
                self._payloads[idx] = payload
                if self._matrix is not None:
                    self._matrix[idx] = vec[0]
                return
            self._ids.append(item_id)
            self._payloads.append(payload)
            if self._matrix is None:
                self._matrix = vec
            else:
                self._matrix = np.vstack([self._matrix, vec])

    def bulk_load(self, items: list[tuple[str, str, dict]]) -> None:
        """Replace the index with a batch of (id, text, payload) tuples."""
        if not items:
            self.reset()
            return
        texts = [t for _, t, _ in items]
        vecs = self._embedder.embed(texts)
        with self._lock:
            self._ids = [i for i, _, _ in items]
            self._payloads = [p for _, _, p in items]
            self._matrix = vecs

    # --- read path --------------------------------------------------------

    def retrieve(self, query: str, k: int = 4) -> list[dict]:
        """
        Return up to k payloads ranked by cosine similarity to the query.
        Each returned dict includes the original payload fields plus a
        `_score` float in [-1, 1].
        """
        if self._matrix is None or len(self._ids) == 0 or not query.strip():
            return []
        qvec = self._embedder.embed([query])[0]  # (dim,)
        scores = self._matrix @ qvec  # cosine, since both are L2-normalized
        top_k = min(k, len(self._ids))
        # argsort ascending → take last `top_k`, reverse for descending.
        top_idx = np.argsort(scores)[-top_k:][::-1]
        out: list[dict] = []
        for i in top_idx:
            entry = dict(self._payloads[int(i)])
            entry["_score"] = float(scores[int(i)])
            out.append(entry)
        return out


# ---------------------------------------------------------------------------
# Population helpers (seed + per-request)
# ---------------------------------------------------------------------------


def populate_from_db(index: RAGIndex, db: dict) -> None:
    """Seed the index from sample_data.json contents."""
    items: list[tuple[str, str, dict]] = []
    for req in db.get("requests", []):
        # Only past *responded* requests have value as precedent.
        if req.get("status") != "Responded" or not req.get("response"):
            continue
        text = _request_text(req)
        items.append(
            (
                f"req:{req['id']}",
                text,
                {
                    "kind": "request",
                    "id": req["id"],
                    "subject": req.get("subject"),
                    "description": req.get("description"),
                    "response": req.get("response"),
                    "status": req.get("status"),
                    "date_filed": req.get("date_filed"),
                },
            )
        )
    for faq in db.get("faqs", []):
        items.append(
            (
                f"faq:{faq['id']}",
                _faq_text(faq),
                {
                    "kind": "faq",
                    "id": faq["id"],
                    "question": faq.get("question"),
                    "answer": faq.get("answer"),
                },
            )
        )
    index.bulk_load(items)
    log.info("RAG index populated with %d items.", len(index))


def index_responded_request(index: RAGIndex, req: dict) -> None:
    """
    Add (or refresh) a responded request in the index. Called when an officer
    approves a request via the admin panel — so each approval grows the corpus
    available to future drafts.
    """
    if req.get("status") != "Responded" or not req.get("response"):
        return
    index.upsert(
        f"req:{req['id']}",
        _request_text(req),
        {
            "kind": "request",
            "id": req["id"],
            "subject": req.get("subject"),
            "description": req.get("description"),
            "response": req.get("response"),
            "status": req.get("status"),
            "date_filed": req.get("date_filed"),
        },
    )


def format_retrieved_for_prompt(hits: list[dict]) -> str:
    """Render retrieved items as a compact, citation-ready block for the LLM prompt."""
    if not hits:
        return "(no related ministry records found in the local archive)"
    lines: list[str] = []
    for i, h in enumerate(hits, start=1):
        if h["kind"] == "request":
            lines.append(
                f"[{i}] Past responded RTI · {h['id']} (filed {h.get('date_filed', '?')})\n"
                f"    Subject: {h.get('subject')}\n"
                f"    Description: {h.get('description')}\n"
                f"    Official response: {h.get('response')}"
            )
        elif h["kind"] == "faq":
            lines.append(
                f"[{i}] FAQ · {h['id']}\n"
                f"    Q: {h.get('question')}\n"
                f"    A: {h.get('answer')}"
            )
    return "\n\n".join(lines)

File: backend-go/internal/rag/rag.go
// Package rag implements a small in-process retrieval layer used to ground
// the AI's drafts in the ministry archive (past responded RTI requests +
// FAQs).
//
// The original Python implementation used sentence-transformers
// (all-MiniLM-L6-v2) for embeddings. Pulling 300 MB of PyTorch into a Go
// binary is impractical, so this Go port uses a deterministic TF-IDF +
// cosine-similarity retriever instead. The retrieval contract is identical:
// callers supply a query string and get back ranked payloads with a `_score`
// in [0, 1].
package rag

import (
	"fmt"
	"math"
	"regexp"
	"sort"
	"strings"
	"sync"

	"github.com/rti4all/backend-go/internal/models"
)

var tokenRe = regexp.MustCompile(`[A-Za-z][A-Za-z0-9]+`)

// stopWords are removed before tokenization to make scoring less noisy.
var stopWords = map[string]struct{}{
	"the": {}, "and": {}, "for": {}, "with": {}, "that": {}, "this": {},
	"from": {}, "are": {}, "was": {}, "were": {}, "have": {}, "has": {},
	"will": {}, "you": {}, "your": {}, "our": {}, "his": {}, "her": {},
	"its": {}, "their": {}, "what": {}, "which": {}, "who": {}, "when": {},
	"where": {}, "why": {}, "how": {}, "any": {}, "all": {}, "more": {},
	"please": {}, "kindly": {}, "would": {}, "could": {}, "should": {},
	"about": {}, "into": {}, "than": {}, "then": {}, "also": {}, "been": {},
	"being": {}, "such": {},
}

func tokenize(text string) []string {
	matches := tokenRe.FindAllString(strings.ToLower(text), -1)
	out := make([]string, 0, len(matches))
	for _, m := range matches {
		if len(m) < 3 {
			continue
		}
		if _, skip := stopWords[m]; skip {
			continue
		}
		out = append(out, m)
	}
	return out
}

type docVector struct {
	terms map[string]float64 // L2-normalized TF-IDF weights
}

// Index is a TF-IDF cosine-similarity store.
type Index struct {
	mu       sync.RWMutex
	ids      []string
	payloads []map[string]any
	vectors  []docVector
	idf      map[string]float64
	docFreq  map[string]int
	docCount int
}

// NewIndex builds an empty index.
func NewIndex() *Index {
	return &Index{idf: map[string]float64{}, docFreq: map[string]int{}}
}

// Len returns the number of indexed items.
func (i *Index) Len() int {
	i.mu.RLock()
	defer i.mu.RUnlock()
	return len(i.ids)
}

func (i *Index) recompute(rawDocs []map[string]int) {
	// Recompute IDF and per-doc TF-IDF vectors from raw term frequencies.
	i.docFreq = map[string]int{}
	for _, tf := range rawDocs {
		for term := range tf {
			i.docFreq[term]++
		}
	}
	i.idf = map[string]float64{}
	n := float64(len(rawDocs))
	for term, df := range i.docFreq {
		// Smoothed inverse document frequency.
		i.idf[term] = math.Log((n+1.0)/(float64(df)+1.0)) + 1.0
	}
	i.docCount = len(rawDocs)
	i.vectors = make([]docVector, len(rawDocs))
	for idx, tf := range rawDocs {
		i.vectors[idx] = i.buildVector(tf)
	}
}

func (i *Index) buildVector(tf map[string]int) docVector {
	v := docVector{terms: make(map[string]float64, len(tf))}
	var norm float64
	for term, count := range tf {
		w := float64(count) * i.idf[term]
		v.terms[term] = w
		norm += w * w
	}
	if norm > 0 {
		norm = math.Sqrt(norm)
		for term := range v.terms {
			v.terms[term] /= norm
		}
	}
	return v
}

// Item is the input form passed to BulkLoad / Upsert.
type Item struct {
	ID      string
	Text    string
	Payload map[string]any
}

// BulkLoad replaces the index contents with the provided items.
func (i *Index) BulkLoad(items []Item) {
	i.mu.Lock()
	defer i.mu.Unlock()

	i.ids = make([]string, 0, len(items))
	i.payloads = make([]map[string]any, 0, len(items))

	rawDocs := make([]map[string]int, 0, len(items))
	for _, it := range items {
		i.ids = append(i.ids, it.ID)
		i.payloads = append(i.payloads, it.Payload)
		tf := map[string]int{}
		for _, tok := range tokenize(it.Text) {
			tf[tok]++
		}
		rawDocs = append(rawDocs, tf)
	}
	i.recompute(rawDocs)
}

// Upsert adds or replaces a single item. It triggers a full IDF recompute
// because adding documents shifts the global term statistics; at the
// expected corpus size (hundreds of items) this is essentially free.
func (i *Index) Upsert(id, text string, payload map[string]any) {
	i.mu.Lock()
	defer i.mu.Unlock()

	tf := map[string]int{}
	for _, tok := range tokenize(text) {
		tf[tok]++
	}

	// Rebuild raw TF maps for the existing corpus so recompute can re-derive
	// IDF and re-vectorize everything with the new term distribution.
	rawDocs := make([]map[string]int, len(i.ids))
	for idx, v := range i.vectors {
		// We don't keep raw TFs around; reverse-engineer counts from the
		// L2-normalized weight by dividing through IDF. This is only used
		// for IDF/df recomputation, which only needs presence/absence of
		// terms and not exact counts, so we can recover well enough by
		// re-tokenising the underlying payload text.
		_ = v
		raw := map[string]int{}
		txt := payloadText(i.payloads[idx])
		for _, tok := range tokenize(txt) {
			raw[tok]++
		}
		rawDocs[idx] = raw
	}

	replaced := false
	for idx, existing := range i.ids {
		if existing == id {
			i.payloads[idx] = payload
			rawDocs[idx] = tf
			replaced = true
			break
		}
	}
	if !replaced {
		i.ids = append(i.ids, id)
		i.payloads = append(i.payloads, payload)
		rawDocs = append(rawDocs, tf)
	}

	i.recompute(rawDocs)
}

// Hit is a single retrieval result. The original payload fields are returned
// alongside a score in [0, 1].
type Hit struct {
	Payload map[string]any
	Score   float64
}

// Retrieve returns up to k payloads ranked by cosine similarity to the query.
func (i *Index) Retrieve(query string, k int) []Hit {
	i.mu.RLock()
	defer i.mu.RUnlock()

	if len(i.ids) == 0 || strings.TrimSpace(query) == "" {
		return nil
	}
	qTF := map[string]int{}
	for _, tok := range tokenize(query) {
		qTF[tok]++
	}
	qVec := i.buildVector(qTF)
	if len(qVec.terms) == 0 {
		return nil
	}
	type scored struct {
		idx   int
		score float64
	}
	scores := make([]scored, 0, len(i.vectors))
	for idx, v := range i.vectors {
		s := cosine(qVec, v)
		if s > 0 {
			scores = append(scores, scored{idx, s})
		}
	}
	sort.Slice(scores, func(a, b int) bool { return scores[a].score > scores[b].score })
	if k > len(scores) {
		k = len(scores)
	}
	out := make([]Hit, 0, k)
	for _, s := range scores[:k] {
		payload := make(map[string]any, len(i.payloads[s.idx])+1)
		for k, v := range i.payloads[s.idx] {
			payload[k] = v
		}
		out = append(out, Hit{Payload: payload, Score: s.score})
	}
	return out
}

func cosine(a, b docVector) float64 {
	// Both vectors are L2-normalised so cosine == dot product.
	if len(a.terms) > len(b.terms) {
		a, b = b, a
	}
	var sum float64
	for term, wa := range a.terms {
		if wb, ok := b.terms[term]; ok {
			sum += wa * wb
		}
	}
	return sum
}

func payloadText(p map[string]any) string {
	switch p["kind"] {
	case "request":
		return strings.Join([]string{
			asString(p["subject"]),
			asString(p["description"]),
			asString(p["response"]),
		}, "\n")
	case "faq":
		return asString(p["question"]) + "\n" + asString(p["answer"])
	}
	return ""
}

func asString(v any) string {
	if v == nil {
		return ""
	}
	s, ok := v.(string)
	if !ok {
		return ""
	}
	return s
}

// PopulateFromDB seeds the index from the on-disk store. Only past
// *responded* RTI requests and FAQs are indexed — pending requests aren't
// useful as precedent.
func PopulateFromDB(idx *Index, db *models.DB) {
	items := make([]Item, 0, len(db.Requests)+len(db.FAQs))
	for _, r := range db.Requests {
		if r.Status != "Responded" || r.Response == "" {
			continue
		}
		text := strings.Join([]string{r.Subject, r.Description, r.Response}, "\n")
		items = append(items, Item{
			ID:   "req:" + r.ID,
			Text: text,
			Payload: map[string]any{
				"kind":        "request",
				"id":          r.ID,
				"subject":     r.Subject,
				"description": r.Description,
				"response":    r.Response,
				"status":      r.Status,
				"date_filed":  r.DateFiled,
			},
		})
	}
	for _, f := range db.FAQs {
		items = append(items, Item{
			ID:   "faq:" + f.ID,
			Text: f.Question + "\n" + f.Answer,
			Payload: map[string]any{
				"kind":     "faq",
				"id":       f.ID,
				"question": f.Question,
				"answer":   f.Answer,
			},
		})
	}
	idx.BulkLoad(items)
}

// IndexResponded adds (or refreshes) one responded request after officer
// approval — closing the feedback loop so future drafts can cite it.
func IndexResponded(idx *Index, r *models.RTIRequest) {
	if r.Status != "Responded" || r.Response == "" {
		return
	}
	text := strings.Join([]string{r.Subject, r.Description, r.Response}, "\n")
	idx.Upsert("req:"+r.ID, text, map[string]any{
		"kind":        "request",
		"id":          r.ID,
		"subject":     r.Subject,
		"description": r.Description,
		"response":    r.Response,
		"status":      r.Status,
		"date_filed":  r.DateFiled,
	})
}

// FormatForPrompt renders retrieval hits as a citation-ready block for the
// AI system prompt.
func FormatForPrompt(hits []Hit) string {
	if len(hits) == 0 {
		return "(no related ministry records found in the local archive)"
	}
	var b strings.Builder
	for i, h := range hits {
		p := h.Payload
		switch p["kind"] {
		case "request":
			fmt.Fprintf(&b, "[%d] Past responded RTI · %s (filed %s)\n    Subject: %s\n    Description: %s\n    Official response: %s",
				i+1, asString(p["id"]), asString(p["date_filed"]),
				asString(p["subject"]), asString(p["description"]), asString(p["response"]))
		case "faq":
			fmt.Fprintf(&b, "[%d] FAQ · %s\n    Q: %s\n    A: %s",
				i+1, asString(p["id"]), asString(p["question"]), asString(p["answer"]))
		}
		if i < len(hits)-1 {
			b.WriteString("\n\n")
		}
	}
	return b.String()
}
